python/ray/data/datasource/file_based_datasource.py

```diff
@@ -237,6 +244,32 @@ def expand_paths(
         return expanded_paths, file_sizes
 
 
+class FastFileMetadataProvider(DefaultFileMetadataProvider):
```

Comment:
I find it weird to have those in a *_datasource.py file. Would it make more sense for all those subclasses of FileMetadataProvider to be placed in filed_meta_provider.py? If not, why?

Reply:
I had considered the same, and I think it would be a good idea to move them all over to `file_metadata_provider.py` as part of this PR.

From the perspective of an end-user, they should be able to use `from ray.data.datasource import FastFileMetadataProvider` regardless of which file we put the class inside of, so this shouldn't change much from their perspective.

From the perspective of a Ray Data maintainer, the current organization largely grew out of organic growth of the code, where one-off utility functions have gradually evolved into more generic/extensible classes over time, and thus follow existing conventions of grouping utility class impls with the datasource that uses them. Thus, all file metadata providers given in `file_based_datasource.py` are meant to be used with file-based datasources, while all file metadata providers in `parquet_datasource.py` are meant to be used specifically with the `ParquetDatasource`.

The primary downside of continuing to preserve this type of grouping is that core dependencies like `file_based_datasource.py` starts to grow too large over time (sitting now at >700 LoC), which makes it harder to grok the important parts at a glance or to understand during a quick top-to-bottom read. 

If we were to move all file metadata provider implementations into `file_metadata_provider.py`, I'd expect it to be just over 300 LoC with all classes organized around the central purpose of providing file metadata. This better normalizes the size of our "large files" and keeps them more focused on a single purpose, so I'm slightly in favor of making this move now if we're in agreement that it's beneficial.

We probably also should do the same with block write path providers in a follow-up PR.

Reply:
I refactored this in the latest commit - let me know what you think.

Reply:
Thanks, I think it makes sense for subclasses of FileMetadataProvider live in file_metadata_provider.py.